File: src/main/java/leets/leenk/domain/feed/application/mapper/FeedMapper.java
package leets.leenk.domain.feed.application.mapper;

import leets.leenk.domain.feed.application.dto.response.*;
import leets.leenk.domain.feed.domain.entity.Comment;
import leets.leenk.domain.feed.domain.entity.Feed;
import leets.leenk.domain.feed.domain.entity.LinkedUser;
import leets.leenk.domain.media.application.dto.response.FeedMediaResponse;
import leets.leenk.domain.media.domain.entity.Media;
import leets.leenk.domain.user.application.mapper.UserProfileMapper;
import leets.leenk.domain.user.domain.entity.User;
import leets.leenk.global.common.dto.PageableMapperUtil;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Slice;
import org.springframework.stereotype.Component;

import java.util.List;
import java.util.Map;

import static leets.leenk.domain.feed.application.util.FeedDescriptionUtil.normalizeDescription;

@Component
@RequiredArgsConstructor
public class FeedMapper {
    private final UserProfileMapper userProfileMapper;

    public FeedListResponse toFeedListResponse(Slice<Feed> slice, Map<Long, List<Media>> mediaMap) {
        List<FeedResponse> responses = toFeedListResponse(slice.getContent(), mediaMap);

        return FeedListResponse.builder()
                .feeds(responses)
                .pageable(PageableMapperUtil.from(slice))
                .build();
    }

    public FeedListResponse toFeedListResponse(User user, Slice<Feed> slice, Map<Long, List<Media>> mediaMap, boolean includeTotalReaction) {
        List<FeedResponse> responses = toFeedListResponse(slice.getContent(), mediaMap);

        return FeedListResponse.builder()
                .totalReactionCount(includeTotalReaction ? user.getTotalReactionCount() : null)
                .feeds(responses)
                .pageable(PageableMapperUtil.from(slice))
                .build();
    }

    private List<FeedResponse> toFeedListResponse(List<Feed> feeds, Map<Long, List<Media>> mediaMap) {
        return feeds.stream()
                .map(feed -> {
                    List<Media> medias = mediaMap.getOrDefault(feed.getId(), List.of());
                    Media thumbnail = medias.stream().findFirst().orElse(null);

                    return toFeedResponse(feed, thumbnail);
                })
                .toList();
    }

    public FeedResponse toFeedResponse(Feed feed, Media thumbNail) {
        return FeedResponse.builder()
                .feedId(feed.getId())
                .author(toFeedAuthorResponse(feed))
                .thumbNail(thumbNail.getThumbnailUrl())
                .totalReactionCount(feed.getTotalReactionCount())
                .build();
    }

    public FeedAuthorResponse toFeedAuthorResponse(Feed feed) {
        return FeedAuthorResponse.builder()
                .author(userProfileMapper.toProfile(feed.getUser()))
                .build();
    }

    public Feed toFeed(User user, String description) {
        String normalizedDescription = normalizeDescription(description);

        return Feed.builder()
                .user(user)
                .description(normalizedDescription)
                .build();
    }

    public FeedDetailResponse toFeedDetailResponse(Feed feed, List<Media> medias, List<LinkedUser> linkedUsers, List<Comment> comments) {
        return FeedDetailResponse.builder()
                .feedId(feed.getId())
                .author(toFeedAuthorResponse(feed))
                .description(feed.getDescription())
                .totalReactionCount(feed.getTotalReactionCount())
                .createdAt(feed.getCreateDate())
                .media(toFeedMediaResponses(medias))
                .linkedUserCount(linkedUsers.size())
                .linkedUser(toLinkedUserResponses(linkedUsers, feed))
                .comments(toGetCommentsResponses(comments))
                .build();
    }

    private List<FeedMediaResponse> toFeedMediaResponses(List<Media> medias) {
        return medias.stream()
                .map(media -> FeedMediaResponse.builder()
                        .position(media.getPosition())
                        .mediaUrl(media.getMediaUrl())
                        .mediaType(media.getMediaType())
                        .build())
                .toList();
    }

    private List<LinkedUserResponse> toLinkedUserResponses(List<LinkedUser> linkedUsers, Feed feed) {
        Long authorId = feed.getUser().getId();

        return linkedUsers.stream()
                .map(linkedUser -> LinkedUserResponse.builder()
                        .user(userProfileMapper.toProfile(linkedUser.getUser()))
                        .isAuthor(linkedUser.getUser().getId().equals(feed.getUser().getId()))
                        .build())
                .toList();
    }

    private List<FeedCommentResponse> toGetCommentsResponses(List<Comment> comments) {
        return comments.stream()
                .map(comment -> FeedCommentResponse.builder()
                        .commentId(comment.getCommentId())
                        .user(userProfileMapper.toProfile(comment.getUser()))
                        .comment(comment.getComment())
                        .createdAt(comment.getCreateDate())
                        .build())
                .toList();
    }

    public FeedNavigationResponse toFeedNavigationResponse(
            Feed currentFeed,
            List<Feed> prevFeeds,
            List<Feed> nextFeeds,
            Map<Long, List<Media>> mediaMap,
            Map<Long, List<LinkedUser>> linkedUserMap,
            Map<Long, List<Comment>> commentsMap,
            boolean hasMorePrev,
            boolean hasMoreNext
    ) {
        FeedDetailResponse current = toFeedDetailResponse(
                currentFeed,
                mediaMap.getOrDefault(currentFeed.getId(), List.of()),
                linkedUserMap.getOrDefault(currentFeed.getId(), List.of()),
                commentsMap.getOrDefault(currentFeed.getId(), List.of())
        );

        List<FeedDetailResponse> prevFeedResponses = prevFeeds.stream()
                .map(feed -> toFeedDetailResponse(
                        feed,
                        mediaMap.getOrDefault(feed.getId(), List.of()),
                        linkedUserMap.getOrDefault(feed.getId(), List.of()),
                        commentsMap.getOrDefault(feed.getId(), List.of())
                ))
                .toList();

        List<FeedDetailResponse> nextFeedResponses = nextFeeds.stream()
                .map(feed -> toFeedDetailResponse(
                        feed,
                        mediaMap.getOrDefault(feed.getId(), List.of()),
                        linkedUserMap.getOrDefault(feed.getId(), List.of()),
                        commentsMap.getOrDefault(feed.getId(), List.of())
                ))
                .toList();

        return FeedNavigationResponse.builder()
                .current(current)
                .prevFeeds(prevFeedResponses)
                .nextFeeds(nextFeedResponses)
                .hasMorePrev(hasMorePrev)
                .hasMoreNext(hasMoreNext)
                .build();
    }
}

File: src/main/kotlin/leets/leenk/domain/feed/application/mapper/FeedMapper.kt
package leets.leenk.domain.feed.application.mapper

import leets.leenk.domain.feed.application.dto.response.FeedAuthorResponse
import leets.leenk.domain.feed.application.dto.response.FeedCommentResponse
import leets.leenk.domain.feed.application.dto.response.FeedDetailResponse
import leets.leenk.domain.feed.application.dto.response.FeedListResponse
import leets.leenk.domain.feed.application.dto.response.FeedNavigationResponse
import leets.leenk.domain.feed.application.dto.response.FeedResponse
import leets.leenk.domain.feed.application.dto.response.LinkedUserResponse
import leets.leenk.domain.feed.application.util.FeedDescriptionUtil.normalizeDescription
import leets.leenk.domain.feed.domain.entity.Comment
import leets.leenk.domain.feed.domain.entity.Feed
import leets.leenk.domain.feed.domain.entity.LinkedUser
import leets.leenk.domain.media.application.dto.response.FeedMediaResponse
import leets.leenk.domain.media.domain.entity.Media
import leets.leenk.domain.user.application.mapper.UserProfileMapper
import leets.leenk.domain.user.domain.entity.User
import leets.leenk.global.common.dto.PageableMapperUtil
import org.springframework.data.domain.Slice
import org.springframework.stereotype.Component

@Component
class FeedMapper(
    private val userProfileMapper: UserProfileMapper,
) {

    fun toFeedListResponse(slice: Slice<Feed>, mediaMap: Map<Long, List<Media>>): FeedListResponse {
        val responses = toFeedListResponse(slice.content, mediaMap)

        return FeedListResponse(
            totalReactionCount = null,
            feeds = responses,
            pageable = PageableMapperUtil.from(slice),
        )
    }

    fun toFeedListResponse(
        user: User,
        slice: Slice<Feed>,
        mediaMap: Map<Long, List<Media>>,
        includeTotalReaction: Boolean,
    ): FeedListResponse {
        val responses = toFeedListResponse(slice.content, mediaMap)

        return FeedListResponse(
            totalReactionCount = if (includeTotalReaction) user.totalReactionCount else null,
            feeds = responses,
            pageable = PageableMapperUtil.from(slice),
        )
    }

    private fun toFeedListResponse(feeds: List<Feed>, mediaMap: Map<Long, List<Media>>): List<FeedResponse> {
        return feeds.map { feed ->
            val medias = mediaMap.getOrDefault(feed.id, emptyList())
            val thumbnail = medias.firstOrNull()

            toFeedResponse(feed, thumbnail)
        }
    }

    fun toFeedResponse(feed: Feed, thumbNail: Media?): FeedResponse {
        return FeedResponse(
            feedId = feed.id!!,
            author = toFeedAuthorResponse(feed),
            thumbNail = thumbNail!!.thumbnailUrl,
            totalReactionCount = feed.totalReactionCount,
        )
    }

    fun toFeedAuthorResponse(feed: Feed): FeedAuthorResponse {
        return FeedAuthorResponse(
            author = userProfileMapper.toProfile(feed.user),
        )
    }

    fun toFeed(user: User, description: String?): Feed {
        val normalizedDescription = normalizeDescription(description)

        return Feed(
            user = user,
            description = normalizedDescription,
        )
    }

    fun toFeedDetailResponse(
        feed: Feed,
        medias: List<Media>,
        linkedUsers: List<LinkedUser>,
        comments: List<Comment>,
    ): FeedDetailResponse {
        return FeedDetailResponse(
            feedId = feed.id!!,
            author = toFeedAuthorResponse(feed),
            description = feed.description,
            totalReactionCount = feed.totalReactionCount,
            createdAt = feed.createDate!!,
            media = toFeedMediaResponses(medias),
            linkedUserCount = linkedUsers.size.toLong(),
            linkedUser = toLinkedUserResponses(linkedUsers, feed),
            comments = toGetCommentsResponses(comments),
        )
    }

    private fun toFeedMediaResponses(medias: List<Media>): List<FeedMediaResponse> {
        return medias.map { media ->
            FeedMediaResponse(
                position = media.position,
                mediaUrl = media.mediaUrl,
                mediaType = media.mediaType,
            )
        }
    }

    private fun toLinkedUserResponses(linkedUsers: List<LinkedUser>, feed: Feed): List<LinkedUserResponse> {
        return linkedUsers.map { linkedUser ->
            LinkedUserResponse(
                user = userProfileMapper.toProfile(linkedUser.user),
                isAuthor = linkedUser.user.id == feed.user.id,
            )
        }
    }

    private fun toGetCommentsResponses(comments: List<Comment>): List<FeedCommentResponse> {
        return comments.map { comment ->
            FeedCommentResponse(
                commentId = comment.commentId!!,
                user = userProfileMapper.toProfile(comment.user),
                comment = comment.comment!!,
                createdAt = comment.createDate!!,
            )
        }
    }

    fun toFeedNavigationResponse(
        currentFeed: Feed,
        prevFeeds: List<Feed>,
        nextFeeds: List<Feed>,
        mediaMap: Map<Long, List<Media>>,
        linkedUserMap: Map<Long, List<LinkedUser>>,
        commentsMap: Map<Long, List<Comment>>,
        hasMorePrev: Boolean,
        hasMoreNext: Boolean,
    ): FeedNavigationResponse {
        val current = toFeedDetailResponse(
            currentFeed,
            mediaMap.getOrDefault(currentFeed.id, emptyList()),
            linkedUserMap.getOrDefault(currentFeed.id, emptyList()),
            commentsMap.getOrDefault(currentFeed.id, emptyList()),
        )

        val prevFeedResponses = prevFeeds.map { feed ->
            toFeedDetailResponse(
                feed,
                mediaMap.getOrDefault(feed.id, emptyList()),
                linkedUserMap.getOrDefault(feed.id, emptyList()),
                commentsMap.getOrDefault(feed.id, emptyList()),
            )
        }

        val nextFeedResponses = nextFeeds.map { feed ->
            toFeedDetailResponse(
                feed,
                mediaMap.getOrDefault(feed.id, emptyList()),
                linkedUserMap.getOrDefault(feed.id, emptyList()),
                commentsMap.getOrDefault(feed.id, emptyList()),
            )
        }

        return FeedNavigationResponse(
            current = current,
            prevFeeds = prevFeedResponses,
            nextFeeds = nextFeedResponses,
            hasMorePrev = hasMorePrev,
            hasMoreNext = hasMoreNext,
        )
    }
}
